### PR TITLE
Add json_array_parser parser and assign_keys transformer

### PR DESCRIPTION
**Pull request by RoeiDimi**
**Description:** <Describe what has changed.>

We are using otel-collector as an infrastructure and receive many types of data from a client. The client's sent data is always a form of json and in one use case the json is a simple headerless jarray and so we need a way to parse it and match headers to each field (something similar to what csv_parser does but also supports types supported in a json format and nested objects)

**Link to tracking Issue:** <Issue number if applicable>
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30321

**Testing:** <Describe what testing was performed and which tests were added.>
* unittests
All the tests found in csv_parser were copied and adjusted adding test scenarios for different types (numbers, booleans, null) as well as a test for parsing a nested object as a part of the jarray
* End to end tests
Used generated traffic on a running otel collector thats using the parser and verified the data is as expected in the end table

**Documentation:** <Describe the documentation added.>
* [json_array_parser.md](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/beacea489ff4ae61c0bac4f477c04748944c9054/pkg/stanza/docs/operators/json_array_parser.md)
* [assign_keys.md](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/beacea489ff4ae61c0bac4f477c04748944c9054/pkg/stanza/docs/operators/assign_keys.md)

### CODE REVIEW

File: pkg/stanza/operator/parser/headerless_jarray/headerless_jarray.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/entry"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper"
+
+	"github.com/valyala/fastjson"
+)
+
+const operatorType = "headerless_jarray_parser"
```
Comment (djaglowski) [inline pkg/stanza/operator/parser/headerless_jarray/headerless_jarray.go:20]: This name is quite verbose. Can we just call it `jarray_parser`?

File: pkg/stanza/operator/parser/headerless_jarray/headerless_jarray.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+		ParserConfig: helper.NewParserConfig(operatorID, operatorType),
+	}
+}
+
+// Config is the configuration of a jarray parser operator.
+type Config struct {
+	helper.ParserConfig `mapstructure:",squash"`
+
+	Header          string `mapstructure:"header"`
+	HeaderDelimiter string `mapstructure:"header_delimiter"`
```
Comment (djaglowski) [inline pkg/stanza/operator/parser/headerless_jarray/headerless_jarray.go:43]: Do we need this parameter? It's certainly relevant in CSV, but perhaps we can assume it in this format?

File: pkg/stanza/docs/operators/headerless_jarray_parser.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
+## `headerless_jarray_parser` operator
+
+The `headerless_jarray_parser` operator parses the string-type field selected by `parse_from` with the given header values.
```
Comment (djaglowski) [inline pkg/stanza/docs/operators/headerless_jarray_parser.md:3]: Thanks for adding a readme too. I think it should clearly describe the expected format, including linking to a specification if possible. When I look for a format called "jarray", I find C# classes and such which are not clearly related. Is this link appropriate? https://json-schema.org/understanding-json-schema/reference/array

Comment (djaglowski) [reply pkg/stanza/docs/operators/headerless_jarray_parser.md:3]: Moreover, should the parser be called `json_array_parser`?

File: pkg/stanza/operator/parser/headerless_jarray/headerless_jarray.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+	HeaderAttribute string `mapstructure:"header_attribute"`
+}
+
+// Build will build a jarray parser operator.
+func (c Config) Build(logger *zap.SugaredLogger) (operator.Operator, error) {
```
Comment (djaglowski) [inline pkg/stanza/operator/parser/headerless_jarray/headerless_jarray.go:48]: I know this code is copied from the csv parser due to being so similar, but reviewing it made me realize several things which I would recommend changing. I've made those changes on https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30645. Do you mind picking them up as well?

File: pkg/stanza/operator/parser/headerless_jarray/headerless_jarray.go (hunk 1 above)
Comment (RoeiDimi) [reply pkg/stanza/operator/parser/headerless_jarray/headerless_jarray.go]: changed to `json_array_parser`

File: pkg/stanza/operator/parser/headerless_jarray/headerless_jarray.go (hunk 2 above)
Comment (RoeiDimi) [reply pkg/stanza/operator/parser/headerless_jarray/headerless_jarray.go]: removed all mentions of a header delimiter, the parser is now using this instead - `const jsonHeaderDelimiter = ","`

File: pkg/stanza/docs/operators/headerless_jarray_parser.md (hunk 3 above)
Comment (RoeiDimi) [reply pkg/stanza/docs/operators/headerless_jarray_parser.md]: yeah i know what you mean about googling it.. Looks like "json array" gives better results and chatgpt did a good job of explaining it to me hehe..

I updated the readme with examples for json arrays and what is expected and also added that link in there.

the renaming to `json_array_parser` is also done

File: pkg/stanza/operator/parser/headerless_jarray/headerless_jarray.go (hunk 4 above)
Comment (RoeiDimi) [reply pkg/stanza/operator/parser/headerless_jarray/headerless_jarray.go]: nice changes, added them as well

File: pkg/stanza/docs/operators/json_array_parser.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+Jarray after parsing:
+```json
+["Hello", 42, {"name": "Alice", "age": 25}, [1, 2, 3], true, null]
+```
+
+Notice that for this example, the current parser will parse every nested object as a string and so the result is actually this - 
+```json
+["Hello", 42, "{\"name\": \"Alice\", \"age\": 25}", "[1, 2, 3]", true, null]
+```
```
[the remark this answers is not in the document]
Comment (djaglowski) [reply pkg/stanza/docs/operators/json_array_parser.md:39]: This has me revisiting [your comment here](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30321#issuecomment-1890911017) about potentially creating a separate `headers_matching_parser`.

I still don't think we should change the csv parser (both for stability reasons and because headers are required in csv format). However, I could see value in this parser without any relation to headers.

Then separately, we could have a transformer which, as you suggested, would map an array into an object by providing a key for each index. This could also be useful in other circumstances as well. I would maybe call it `assign_keys` and config would look something like this:

```yaml
# sample config
- type: assign_keys
  keys: [ "foo", "bar" ]
  field: body

# example input
[ 123, "hello" ]

# expected output
{ "foo": 123, "bar": "hello" }
```

What do you think? I know we've iterated on this design quite a bit but I want to get it right and both of these operators could be quite simple.

File: pkg/stanza/docs/operators/json_array_parser.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+A JArray string (or a json array string) is a string that represents a JSON array. A JSON array is a type of data structure that is used to store data in a structured way. It consists of an ordered list of values that can be either strings, numbers, objects, or even other arrays.
+#### Examples:
+a simple Jarray string with strictly strings in it:
+```
+"[\"Ford\", \"BMW\", \"Fiat\"]"
+```
+
+Jarray after parsing:
+```json
+["Ford", "BMW", "Fiat"]
```
[the remark this answers is not in the document]
Comment (RoeiDimi) [reply pkg/stanza/docs/operators/json_array_parser.md]: changed to something meaningless :)

File: pkg/stanza/docs/operators/json_array_parser.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+|--------------------|------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------|
+| `id`               | `json_array_parser`                             | A unique identifier for the operator.                                                                                                             |
+| `output`           | Next in pipeline                         | The connected operator(s) that will receive all outbound entries.                                                                                 |
+| `header`           | required when `header_attribute` not set | A string of delimited field names                                                                                                                 |
+| `header_attribute` | required when `header` not set           | An attribute name to read the header field from, to support dynamic field names                                                                   |
```
[the remark this answers is not in the document]
Comment (RoeiDimi) [reply pkg/stanza/docs/operators/json_array_parser.md]: Done. a new `assign_keys` transformer introduced and now `json_array_parser ` does only the part of parsing a jarray string into a list of objects.

File: pkg/stanza/operator/parser/json_array/json_array_parser.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+		pp:             pp,
+		parse:          generateJarrayParseFunc(pp),
+	}, nil
+}
+
+// Parser is an operator that parses jarray in an entry.
+type Parser struct {
+	helper.ParserOperator
+	parse parseFunc
+	pp    *fastjson.ParserPool
```
Comment (djaglowski) [inline pkg/stanza/operator/parser/json_array/json_array_parser.go:61]: nit: can we call this `pool`?
```suggestion
	pool    *fastjson.ParserPool
```

File: pkg/stanza/operator/parser/json_array/json_array_parser.go [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+func (c Config) Build(logger *zap.SugaredLogger) (operator.Operator, error) {
+	parserOperator, err := c.ParserConfig.Build(logger)
+	if err != nil {
+		return nil, err
+	}
+
+	pp := &fastjson.ParserPool{}
+	return &Parser{
+		ParserOperator: parserOperator,
+		pp:             pp,
```
Comment (djaglowski) [inline pkg/stanza/operator/parser/json_array/json_array_parser.go:52]: nit
```suggestion
	return &Parser{
		ParserOperator: parserOperator,
		pp:             new(fastjson.ParserPool),
```

File: pkg/stanza/operator/parser/json_array/json_array_parser.go [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+	"github.com/valyala/fastjson"
+)
+
+const operatorType = "json_array_parser"
+
+func init() {
+	operator.Register(operatorType, func() operator.Builder { return NewConfig() })
+}
+
+// NewConfig creates a new jarray parser config with default values
```
Comment (djaglowski) [inline pkg/stanza/operator/parser/json_array/json_array_parser.go:25]: nit (please update the other comments as well)
```suggestion
// NewConfig creates a new json_array parser config with default values
```

File: pkg/stanza/operator/parser/json_array/json_array_parser.go [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+// Parser is an operator that parses jarray in an entry.
+type Parser struct {
+	helper.ParserOperator
+	parse parseFunc
+	pp    *fastjson.ParserPool
+}
+
+type parseFunc func(any) (any, error)
```
Comment (djaglowski) [inline pkg/stanza/operator/parser/json_array/json_array_parser.go:64]: We should be able to remove this. It was introduced into the csv parser because in some cases we needed to dynamically generate a parse func. However, with our design changes, we should be able to use a static function in this parser.

File: pkg/stanza/operator/parser/json_array/json_array_parser.go [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+// Process will parse an entry for jarray.
+func (r *Parser) Process(ctx context.Context, e *entry.Entry) error {
+	return r.ParserOperator.ProcessWith(ctx, e, r.parse)
+}
+
+func generateJarrayParseFunc(pp *fastjson.ParserPool) parseFunc {
+	return func(value any) (any, error) {
```
Comment (djaglowski) [inline pkg/stanza/operator/parser/json_array/json_array_parser.go:72]: Likewise, the returned function can just be a receiver method on the parser itself.

File: pkg/stanza/operator/parser/json_array/json_array_parser.go [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper"
+
+	"github.com/valyala/fastjson"
+)
+
+const operatorType = "json_array_parser"
+
+func init() {
+	operator.Register(operatorType, func() operator.Builder { return NewConfig() })
```
Comment (djaglowski) [inline pkg/stanza/operator/parser/json_array/json_array_parser.go:22]: I think unfortunately we need to add these new operators behind a feature gate [(example)](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/82484a272ed2f1d324021c1518b5cbf2c82d9d43/pkg/stanza/fileconsumer/matcher/matcher.go#L119). Otherwise, we're immediately adding them to the filelog receiver, which means no breaking changes to the config. Hopefully we can move the gate forward quickly, but let's at least have one release where it has to be opted into.

File: pkg/stanza/operator/transformer/assign_keys/assign_keys.go [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/errors"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper"
+	"go.uber.org/zap"
+)
+
+const operatorType = "assign_keys"
+
+func init() {
+	operator.Register(operatorType, func() operator.Builder { return NewConfig() })
```
Comment (djaglowski) [inline pkg/stanza/operator/transformer/assign_keys/assign_keys.go:17]: We'll need a feature gate here as well

File: pkg/stanza/operator/transformer/assign_keys/assign_keys.go [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+// Process will process an entry with AssignKeys transformation.
+func (p *Transformer[T]) Process(ctx context.Context, entry *entry.Entry) error {
+	return p.ProcessWith(ctx, entry, p.Transform)
+}
+
+// Transform will apply AssignKeys to an entry
+func (p *Transformer[T]) Transform(entry *entry.Entry) error {
+	inputListInterface, ok := entry.Delete(p.Field)
```
Comment (djaglowski) [inline pkg/stanza/operator/transformer/assign_keys/assign_keys.go:94]: We follow a pattern where parsing and transformation operations should not be destructive, by default. It is a little more verbose for the user, but the original field should be left in place. If the user chooses to `parse_to` the same field as `parse_from`, it will naturally overwrite it. Otherwise, they can delete the field explicitly with the `remove` operator.

This means we should use `Get` here and shouldn't need any rollback behavior.

File: pkg/stanza/operator/transformer/assign_keys/assign_keys.go [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+	return nil, fmt.Errorf("invalid field type: %T", c.Field.FieldInterface)
+}
+
+// Transformer transforms a list in the entry field into a map. Each value is assigned a key from configuration keys
+type Transformer[T interface {
+	entry.BodyField | entry.ResourceField | entry.AttributeField
+	entry.FieldInterface
+}] struct {
+	helper.TransformerOperator
+	Field T
```
Comment (djaglowski) [inline pkg/stanza/operator/transformer/assign_keys/assign_keys.go:83]: Why not just `entry.Field`?

File: pkg/stanza/operator/transformer/assign_keys/assign_keys.go [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+		// The field doesn't exist, so ignore it
+		return fmt.Errorf("apply assign_keys: field %s does not exist on entry", p.Field)
+	}
+
+	inputList, ok := inputListInterface.([]any)
+	if !ok {
+		return p.rollbackTransformAndGetError(entry, inputListInterface, fmt.Sprintf("apply assign_keys: couldn't convert field %s to []any", p.Field))
+	}
+	if len(inputList) != len(p.Keys) {
+		return p.rollbackTransformAndGetError(entry, inputListInterface, fmt.Sprintf("apply assign_keys: field %s contains %d values while expected keys are %s contain %d keys", p.Field, len(inputList), p.Keys, len(p.Keys)))
```
Comment (djaglowski) [inline pkg/stanza/operator/transformer/assign_keys/assign_keys.go:105]: I think we should document what happens when the number of fields doesn't match the number of keys.

**Review by djaglowski — COMMENTED** [review]
Thanks @RoeiDimi, the design feels right. I left a lot of comments but mostly just nits. A few minor things I think we need to take care of though.

File: pkg/stanza/docs/operators/assign_keys.md [hunk 18]
```diff
@@ -0,0 +1,99 @@
+## `assign_keys` operator
```
[the remark this answers is not in the document]
Comment (RoeiDimi) [reply pkg/stanza/docs/operators/assign_keys.md:1]: Done

File: pkg/stanza/operator/parser/json_array/json_array_parser.go (hunk 8 above)
Comment (RoeiDimi) [reply pkg/stanza/operator/parser/json_array/json_array_parser.go]: Done

File: pkg/stanza/operator/parser/json_array/json_array_parser.go (hunk 9 above)
Comment (RoeiDimi) [reply pkg/stanza/operator/parser/json_array/json_array_parser.go]: Done

File: pkg/stanza/operator/parser/json_array/json_array_parser.go (hunk 10 above)
Comment (RoeiDimi) [reply pkg/stanza/operator/parser/json_array/json_array_parser.go]: Done

File: pkg/stanza/operator/parser/json_array/json_array_parser.go (hunk 12 above)
Comment (RoeiDimi) [reply pkg/stanza/operator/parser/json_array/json_array_parser.go]: Done

File: pkg/stanza/operator/parser/json_array/json_array_parser.go (hunk 13 above)
Comment (RoeiDimi) [reply pkg/stanza/operator/parser/json_array/json_array_parser.go]: Done but i wasn't sure about the flag name and usage.. I used `filelog.jsonParserArray` and `filelog.assignKeys`.
let me know if I used it correctly.

File: pkg/stanza/operator/transformer/assign_keys/assign_keys.go (hunk 15 above)
Comment (RoeiDimi) [reply pkg/stanza/operator/transformer/assign_keys/assign_keys.go]: Done

File: pkg/stanza/operator/transformer/assign_keys/assign_keys.go (hunk 16 above)
Comment (RoeiDimi) [reply pkg/stanza/operator/transformer/assign_keys/assign_keys.go]: Changed

File: pkg/stanza/docs/operators/assign_keys.md [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+The `assign_keys` assigns keys from the configuration to an input list. the output is a map containing these key-value pairs
+
+### Configuration Fields
+
+| Field      | Default          | Description |
+| ---        | ---              | ---         |
+| `id`       | `assign_keys`        | A unique identifier for the operator. |
+| `output`   | Next in pipeline | The connected operator(s) that will receive all outbound entries. |
+| `field`    | required         | The [field](../types/field.md) to assign keys to. |
+| `keys`    | required         | The list of strings to be used as the keys to the input list's values. It's length is expected to be equal to the length of the values list from field. In case there is a mismatch, assign_keys will return an error and the field will remain untouched|
```
Comment (djaglowski) [inline pkg/stanza/docs/operators/assign_keys.md:12]: Thanks for adding the note, just compressing this a bit.
```suggestion
| `keys`    | required         | The list of strings to be used as the keys to the input list's values. Its length is expected to be equal to the length of the values list from field. In case there is a mismatch, an error will result. |
```

File: pkg/stanza/operator/parser/json_array/json_array_parser.go [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper"
+
+	"github.com/valyala/fastjson"
+)
+
+const operatorType = "json_array_parser"
+
+var jsonArrayParserFeatureGate = featuregate.GlobalRegistry().MustRegister(
+	"filelog.jsonParserArray",
```
Comment (djaglowski) [inline pkg/stanza/operator/parser/json_array/json_array_parser.go:23]: Technically this is available for more receivers as well. Also not sure the best name but I think this is sufficient.
```suggestion
	"logs.jsonParserArray",
```

File: pkg/stanza/operator/transformer/assign_keys/assign_keys.go [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper"
+	"go.opentelemetry.io/collector/featuregate"
+	"go.uber.org/zap"
+)
+
+const operatorType = "assign_keys"
+
+var assignKeysTransformerFeatureGate = featuregate.GlobalRegistry().MustRegister(
+	"filelog.assignKeys",
```
Comment (djaglowski) [inline pkg/stanza/operator/transformer/assign_keys/assign_keys.go:17]:
```suggestion
	"logs.assignKeys",
```

File: pkg/stanza/operator/transformer/assign_keys/testdata/config.yaml [hunk 22]
```diff
@@ -0,0 +1,12 @@
...
+  field: body
+  keys: ["foo", "bar"]
+assign_keys_body_nested:
+  type: assign_keys
+  field: body.nested
+  keys: ["foo", "bar"]
+assign_keys_attributes:
+  type: assign_keys
+  field: attributes.input
+  keys: ["foo", "bar"]
```
Comment (djaglowski) [inline pkg/stanza/operator/transformer/assign_keys/testdata/config.yaml:12]:
```suggestion
  keys: ["foo", "bar"]

```

**Review by djaglowski — APPROVED** [review]

**Comment by RoeiDimi** [comment]
Thanks for fixing and approving @djaglowski .
Do the build errors make sense to you? it looks like it has an issue importing the fastjson package I used (locally i run the tests etc and it looks ok)
Also, is there anything else i need to do before this is merged?

**Comment by djaglowski** [comment]
@RoeiDimi, try running `make goporto`, `make crosslink`, and `make -j8 gotidy`.

**Comment by djaglowski** [comment]
Hmm, not sure if there are problems specific to Windows here but I know we don't have a lot of contributors on that OS so it may be. 

I pulled down your branch and found a few lint issues. Most notably we can't have underscores in the package names (but can leave them in the operator types).

I haven't figured out the fastjson problem yet but I'll push an update with some fixes I'm working through either way.

**Comment by djaglowski** [comment]
@RoeiDimi, I was able to resolve the issue by adding the new dependency directly to the root module. Thanks for the new features and for iterating on the design with me.

**Comment by RoeiDimi** [comment]
Thanks @djaglowski, i'ts been a pleasure working together!

One last question though - assign_keys can't really parse the values straight into attributes (the way csv_parser can)
can you think of an easy way to make that happen?

i mean, I now understand that all this may not solve my initial issue completely.
with the initial design i could do this - 
```
- type: csv_parser
      header: TimeGenerated,SourceIP,SourcePort,DestinationIP,DestinationPort,Protocol,SentBytes,ReceivedBytes,ExtID
      parse_from: body
      parse_to: attributes
      is_jarray: true
```
we have a generic exporter code that reads all the key-values from attributes and sends them.

Can you think of a way I could use the json_array_parser -> assign_keys design to achieve that?

**Comment by djaglowski** [comment]
@RoeiDimi, looks like we lost track of the goal somewhat but luckily you should be able to flatten after assigning keys:
```yaml
    operators:
      - type: json_array_parser
        parse_from: body
        parse_to: attributes.tmp
      - type: assign_keys
        keys: ["a", "b", "c"]
        field: attributes.tmp
      - type: flatten
        field: attributes.tmp
```

**Comment by RoeiDimi** [comment]
@djaglowski yeah that's pretty cool, thank you so much! :)

I apologize but I have another question :\
About the feature gate we added - how do I enable it? I saw [this link](https://github.com/open-telemetry/opentelemetry-collector/blob/main/featuregate/README.md)

so I tried running the collector with all the following variations but none worked - 

- `--feature-gates=logs.assignKeys,logs.jsonParserArray`
- `--feature-gates=assignKeys,+jsonParserArray`
- `--feature-gates=assignKeys,jsonParserArray`
- `--feature-gates=+logs.assignKeys,+logs.jsonParserArray`

**Comment by sumo-drosiek** [comment]
@djaglowski, @RoeiDimi  Shouldn't we have the functionality in OTTL language? I was able to achive the same functionality I believe using `transformprocessor:

```
exporters:
  logging:
    verbosity: detailed
processors:
  transform/parsejson:
    log_statements:
      - context: log
        statements:
          // ParseJSON expects `{}` so I need to build it manually
          - set(body, ParseJSON(Concat(["{\"tmp\":", body, "}"], "")))
          - set(body["key1"], body["tmp"][0])
          - set(body["key2"], body["tmp"][1])
          - set(body["key3"], body["tmp"][2])
          - delete_key(body, "tmp")
receivers:
  filelog/containers:
    start_at: beginning
    include:
      - ./tmp/logs/5.json
service:
  pipelines:
    logs/containers:
      exporters:
      - logging
      processors:
      - transform/parsejson
      receivers:
      - filelog/containers

```

./tmp/logs/5.json

```
["value1", "value2", "value3"]
```

output:

```
LogRecord #0
ObservedTimestamp: 2024-01-24 08:11:49.193363736 +0000 UTC
Timestamp: 1970-01-01 00:00:00 +0000 UTC
SeverityText: 
SeverityNumber: Unspecified(0)
Body: Map({"key1":"value1","key2":"value2","key3":"value3"})
Attributes:
     -> log.file.name: Str(5.json)
Trace ID: 
Span ID: 
Flags: 0
        {"kind": "exporter", "data_type": "logs", "name": "logging"}
```

**Comment by RoeiDimi** [comment]
@sumo-drosiek @djaglowski , this suggestion may be interesting as it seems that using the 3 operators together (json_array_parser, assign_keys, flatten) result in a massive performance hit. Seems like operators are not very efficient in general as the initial design with a single operator (headerless_jarray_parser) that does the logic in json_array_parser+assign_keys was significantly better performance-wise.

anyway @sumo-drosiek I'd like to test your suggestion, i tried adding this to the builder config - 
```
processors:
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/transformprocessor v0.90.0
```

but im getting and error saying the revision is wrong - 
`go: github.com/open-telemetry/opentelemetry-collector-contrib/transformprocessor@v0.90.0: reading github.com/open-telemetry/opentelemetry-collector-contrib/transformprocessor/go.mod at revision transformprocessor/v0.90.0: unknown revision transformprocessor/v0.90.0`

is it not `v0.90.0`? can you please share the revision you used?

Thanks guys for helping me through this entirely new world of OTEL collector hehe..

**Comment by RoeiDimi** [comment]
Hi @djaglowski 
I know this is closed as completed but after trying to use the solution we went with, i see too big of a performance hit to be able to use it. Using these 3 operators is 2X slower than using the original `headerless_jarray_parser ` alone.
I also tried @sumo-drosiek 's suggestion (thank you for trying to help!) and it shows roughly the same numbers as using the 3 operators

For my benchmarks I'm using Azure kubernetes, single node, single pod. I tested the following:

1. `headerless-jarray-parser` can take **~10-11 mil** events per minute
2. `3 operators chained` (or the processor suggestion)  **~4-5 mil** events per minute
3. Just for testing i removed the flatten to see if this is the only issue but its just a part of it, i got **~6-7 mil** events per minute
4. `json_array_parser` alone got me back to the **~10-11** mil events per minute

It also makes sense in a way as `headerless-jarray-parser` could iterate only once on the input array, create a map and then replace attributes with that map..

I totally get the refactor we went for from a modularity point of view but is there any chance you'd agree to now also adding `headerless-jarray-parser` for performance reasons?

**Comment by RoeiDimi** [comment]
@djaglowski yes, that's a good, clean idea
I guess we can now continue here - https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30814
:)
